// d2-utilizr/src/index.js
import isEmpty from './isEmpty';
import isArray from './isArray';
import isObject from './isObject';
import isString from './isString';
import isNumber from './isNumber';

// Value checks
export {default as isEmpty} from './isEmpty';
export {default as isDefined} from './isDefined';

// Type checks
export {default as isObject} from './isObject';
export {default as isFunction} from './isFunction';
export {default as isBoolean} from './isBoolean';
export {default as isPrimitive} from './isPrimitive';
export {default as isNumber} from './isNumber';
export {default as isString} from './isString';
export {default as isArray} from './isArray';
export {default as isNumeric} from './isNumeric';
export {default as isInteger} from './isInteger';

// Number functions
export {default as numberConstrain} from './numberConstrain';
export {default as numberToFixed} from './numberToFixed';
export {default as numberIsFinite} from './numberIsFinite';
export {default as numberDecimals} from './numberDecimals';

// String functions
export {default as stringReplaceAll} from './stringReplaceAll';
export {default as stringTrim} from './stringTrim';
export {default as camelCaseToUnderscores} from './camelCaseToUnderscores';

// Array functions
export {default as arrayPluck} from './arrayPluck';
export {default as arrayContains} from './arrayContains';
export {default as arrayUnique} from './arrayUnique';
export {default as arrayFrom} from './arrayFrom';
export {default as arrayTo} from './arrayTo';
export {default as arrayClean} from './arrayClean';
export {default as arraySort} from './arraySort';
export {default as arrayReplace} from './arrayReplace';
export {default as arrayRepeat} from './arrayRepeat';
export {default as arrayInsert} from './arrayInsert';
export {default as arrayDifference} from './arrayDifference';
export {default as arrayMin} from './arrayMin';
export {default as arrayMax} from './arrayMax';
export {default as isArrayOfStrings} from './isArrayOfStrings';

// Object function
export {default as objectClean} from './objectClean';

// Misc
export {default as isIE} from './isIE';
export {default as clone} from './clone';
export {default as uuid} from './uuid';
export {default as isIterable} from './isIterable';
