// @hisptz/ngx-dhis2-menu/esm5/lib/ngx-dhis2-menu.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { containers } from './containers/index';
import { components } from './components/index';
import { directives } from './directives/index';
import { pipes } from './pipes/index';
var NgxDhis2MenuModule = /** @class */ (function () {
    function NgxDhis2MenuModule() {
    }
    /**
     * @return {?}
     */
    NgxDhis2MenuModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgxDhis2MenuModule,
            providers: []
        };
    };
    NgxDhis2MenuModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, FormsModule, HttpClientModule],
                    declarations: tslib_1.__spread(containers, components, directives, pipes),
                    exports: tslib_1.__spread(containers),
                },] },
    ];
    return NgxDhis2MenuModule;
}());
export { NgxDhis2MenuModule };
