// @hisptz/ngx-dhis2-dictionary/esm2015/lib/components/dictionary-list/dictionary-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { Store } from '@ngrx/store';
import { getDictionaryList } from '../../store/selectors/dictionary.selectors';
import { InitializeDictionaryMetadataAction } from '../../store/actions/dictionary.actions';
export class DictionaryListComponent {
    /**
     * @param {?} store
     */
    constructor(store) {
        this.store = store;
        this.activeItem = 0;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.metadataIdentifiers.length > 0) {
            this.store.dispatch(new InitializeDictionaryMetadataAction(this.metadataIdentifiers));
            this.dictionaryList$ = this.store.select(getDictionaryList(this.metadataIdentifiers));
        }
    }
    /**
     * @param {?} index
     * @param {?} e
     * @return {?}
     */
    setActiveItem(index, e) {
        e.stopPropagation();
        if (this.activeItem === index) {
            this.activeItem = -1;
        }
        else {
            this.activeItem = index;
        }
    }
}
DictionaryListComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'ngx-dhis2-dictionary-list',
                template: `<div class="dictionary-list-container">
  <div
    class="dictionary-item-container"
    *ngFor="let dictionaryItem of dictionaryList$ | async; let i = index"
  >
    <div
      class="dictionary-title"
      (click)="setActiveItem(i, $event)"
    >
      <span>{{ dictionaryItem.name }}</span>
      <div *ngIf="!dictionaryItem.name">
        <app-dictionary-progress></app-dictionary-progress>
      </div>
    </div>
    <div
      class="dictionary-body"
      *ngIf="activeItem === i"
    >
      <div class="dictionary-description" [innerHtml]="dictionaryItem.description"></div>
      <div *ngIf="dictionaryItem.progress.loading">
        <app-dictionary-progress></app-dictionary-progress>
      </div>
    </div>
  </div>
</div>
`,
                styles: [`.dictionary-list-container{box-shadow:0 0 3px rgba(0,0,0,.3)}.dictionary-item-container{border-bottom:thin solid #ececec}.dictionary-title{padding:5px 10px;cursor:pointer}.dictionary-item-container:last-child{border-bottom-style:none}.dictionary-body{border-top:thin solid #ececec;padding:10px}.dictionary-description{text-align:justify}`],
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
DictionaryListComponent.ctorParameters = () => [
    { type: Store }
];
DictionaryListComponent.propDecorators = {
    metadataIdentifiers: [{ type: Input }]
};
function DictionaryListComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DictionaryListComponent.prototype.metadataIdentifiers;
    /** @type {?} */
    DictionaryListComponent.prototype.dictionaryList$;
    /** @type {?} */
    DictionaryListComponent.prototype.activeItem;
    /** @type {?} */
    DictionaryListComponent.prototype.store;
}
