// @hisptz/ngx-dhis2-menu/esm2015/lib/ngx-dhis2-menu.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { containers } from './containers/index';
import { components } from './components/index';
import { directives } from './directives/index';
import { pipes } from './pipes/index';
export class NgxDhis2MenuModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: NgxDhis2MenuModule,
            providers: []
        };
    }
}
NgxDhis2MenuModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, FormsModule, HttpClientModule],
                declarations: [
                    ...containers,
                    ...components,
                    ...directives,
                    ...pipes
                ],
                exports: [...containers],
            },] },
];
