// @hisptz/ngx-dhis2-table/esm2015/lib/helpers/get-table-configuration.helper.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @param {?} favoriteObject
 * @param {?} visualizationLayout
 * @param {?} type
 * @return {?}
 */
export function getTableConfiguration(favoriteObject, visualizationLayout, type) {
    return {
        title: favoriteObject.hasOwnProperty('displayName')
            ? favoriteObject.displayName
            : favoriteObject.hasOwnProperty('name')
                ? favoriteObject.name
                : '',
        subtitle: favoriteObject.hasOwnProperty('subtitle')
            ? favoriteObject.subtitle
            : '',
        showColumnTotal: favoriteObject.hasOwnProperty('colTotal')
            ? favoriteObject.colTotal
            : true,
        showColumnSubtotal: favoriteObject.hasOwnProperty('colSubtotal')
            ? favoriteObject.colSubtotal
            : true,
        showRowTotal: favoriteObject.hasOwnProperty('rowTotal')
            ? favoriteObject.rowTotal
            : true,
        showRowSubtotal: favoriteObject.hasOwnProperty('rowSubtotal')
            ? favoriteObject.rowSubtotal
            : true,
        showDimensionLabels: favoriteObject.hasOwnProperty('showDimensionLabels')
            ? favoriteObject.showDimensionLabels
            : true,
        hideEmptyRows: favoriteObject.hasOwnProperty('hideEmptyRows')
            ? favoriteObject.hideEmptyRows
            : true,
        showHierarchy: favoriteObject.hasOwnProperty('showHierarchy')
            ? favoriteObject.showHierarchy
            : true,
        displayList: checkForEventDataType(favoriteObject, type),
        rows: visualizationLayout.rows ? visualizationLayout.rows : ['pe'],
        columns: visualizationLayout.columns ? visualizationLayout.columns : ['dx'],
        legendSet: favoriteObject.legendSet || null,
        styles: null
    };
}
/**
 * @param {?} favoriteObject
 * @param {?} favoriteType
 * @return {?}
 */
function checkForEventDataType(favoriteObject, favoriteType) {
    let /** @type {?} */ displayList = false;
    if (favoriteType === 'EVENT_REPORT') {
        if (favoriteObject.hasOwnProperty('dataType') &&
            favoriteObject.dataType === 'EVENTS') {
            displayList = true;
        }
    }
    return displayList;
}
