// @hisptz/ngx-dhis2-table/esm5/lib/components/table-list/table-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { getTableConfiguration } from '../../helpers/index';
var TableListComponent = /** @class */ (function () {
    function TableListComponent() {
        this.tableLayers = [];
    }
    /**
     * @return {?}
     */
    TableListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.visualizationLayers && this.visualizationLayers.length > 0) {
            this.tableLayers = this.visualizationLayers.map(function (layer) {
                return {
                    tableConfiguration: getTableConfiguration(layer.config || {}, layer.layout, _this.visualizationType),
                    analyticsObject: layer.analytics
                };
            });
        }
    };
    TableListComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'ngx-dhis2-table-list',
                    template: "<div class=\"row\">\n    <div class=\"col-12\" *ngFor=\"let tableLayer of tableLayers\">\n        <ngx-dhis2-table-item [analyticsObject]=\"tableLayer.analyticsObject\" [tableConfiguration]=\"tableLayer.tableConfiguration\">\n        </ngx-dhis2-table-item>\n    </div>\n</div>",
                    styles: [""]
                },] },
    ];
    /** @nocollapse */
    TableListComponent.ctorParameters = function () { return []; };
    TableListComponent.propDecorators = {
        visualizationLayers: [{ type: Input }],
        visualizationType: [{ type: Input }]
    };
    return TableListComponent;
}());
export { TableListComponent };
function TableListComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    TableListComponent.prototype.visualizationLayers;
    /** @type {?} */
    TableListComponent.prototype.visualizationType;
    /** @type {?} */
    TableListComponent.prototype.tableLayers;
}
