// @hisptz/ngx-dhis2-dictionary/fesm2015/hisptz-ngx-dhis2-dictionary.js
import { createEntityAdapter } from '@ngrx/entity';
import { createFeatureSelector, createSelector, Store, StoreModule } from '@ngrx/store';
import { filter, find, map } from 'lodash';
import { ChangeDetectionStrategy, Component, Input, Injectable, NgModule } from '@angular/core';
import { __decorate, __metadata } from 'tslib';
import { DatePipe, CommonModule } from '@angular/common';
import { Actions, Effect, ofType, EffectsModule } from '@ngrx/effects';
import { mergeMap, map as map$1, tap } from 'rxjs/operators';
import { Observable, from, forkJoin } from 'rxjs';
import { NgxDhis2HttpClientService } from '@hisptz/ngx-dhis2-http-client';
import { HttpClientModule } from '@angular/common/http';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {string} */
const DictionaryActionTypes = {
    InitializeDictionaryMetadata: '[Dictionary] initialize incoming metadata',
    AddDictionaryMetadataList: '[Dictionary] add dictionary metadata list',
    UpdateDictionaryMetadata: '[Dictionary] update dictionary metadata list',
};
class InitializeDictionaryMetadataAction {
    /**
     * @param {?} dictionaryMetadataIdentifiers
     */
    constructor(dictionaryMetadataIdentifiers) {
        this.dictionaryMetadataIdentifiers = dictionaryMetadataIdentifiers;
        this.type = DictionaryActionTypes.InitializeDictionaryMetadata;
    }
}
class AddDictionaryMetadataListAction {
    /**
     * @param {?} dictionaryMetadataList
     */
    constructor(dictionaryMetadataList) {
        this.dictionaryMetadataList = dictionaryMetadataList;
        this.type = DictionaryActionTypes.AddDictionaryMetadataList;
    }
}
class UpdateDictionaryMetadataAction {
    /**
     * @param {?} dictionaryMetadataId
     * @param {?} changes
     */
    constructor(dictionaryMetadataId, changes) {
        this.dictionaryMetadataId = dictionaryMetadataId;
        this.changes = changes;
        this.type = DictionaryActionTypes.UpdateDictionaryMetadata;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const /** @type {?} */ dictionaryAdapter = createEntityAdapter();
const /** @type {?} */ initialState = dictionaryAdapter.getInitialState({});
/**
 * @param {?=} state
 * @param {?=} action
 * @return {?}
 */
function dictionaryReducer(state = initialState, action) {
    switch (action.type) {
        case DictionaryActionTypes.AddDictionaryMetadataList: {
            return dictionaryAdapter.addMany(action.dictionaryMetadataList, state);
        }
        case DictionaryActionTypes.UpdateDictionaryMetadata: {
            return dictionaryAdapter.updateOne({ id: action.dictionaryMetadataId, changes: action.changes }, state);
        }
    }
    return state;
}
const /** @type {?} */ getDictionaryState = createFeatureSelector('dictionary');
const { selectEntities: getDictionaryMetadataEntities } = dictionaryAdapter.getSelectors(getDictionaryState);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const /** @type {?} */ getDictionaryList = metadataIdentifiers => createSelector(getDictionaryMetadataEntities, (dictionaryMetadataEntities) => filter(map(metadataIdentifiers, metadataId => dictionaryMetadataEntities[metadataId]), metadata => metadata));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class DictionaryListComponent {
    /**
     * @param {?} store
     */
    constructor(store) {
        this.store = store;
        this.activeItem = 0;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.metadataIdentifiers.length > 0) {
            this.store.dispatch(new InitializeDictionaryMetadataAction(this.metadataIdentifiers));
            this.dictionaryList$ = this.store.select(getDictionaryList(this.metadataIdentifiers));
        }
    }
    /**
     * @param {?} index
     * @param {?} e
     * @return {?}
     */
    setActiveItem(index, e) {
        e.stopPropagation();
        if (this.activeItem === index) {
            this.activeItem = -1;
        }
        else {
            this.activeItem = index;
        }
    }
}
DictionaryListComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'ngx-dhis2-dictionary-list',
                template: `<div class="dictionary-list-container">
  <div
    class="dictionary-item-container"
    *ngFor="let dictionaryItem of dictionaryList$ | async; let i = index"
  >
    <div
      class="dictionary-title"
      (click)="setActiveItem(i, $event)"
    >
      <span>{{ dictionaryItem.name }}</span>
      <div *ngIf="!dictionaryItem.name">
        <app-dictionary-progress></app-dictionary-progress>
      </div>
    </div>
    <div
      class="dictionary-body"
      *ngIf="activeItem === i"
    >
      <div class="dictionary-description" [innerHtml]="dictionaryItem.description"></div>
      <div *ngIf="dictionaryItem.progress.loading">
        <app-dictionary-progress></app-dictionary-progress>
      </div>
    </div>
  </div>
</div>
`,
                styles: [`.dictionary-list-container{box-shadow:0 0 3px rgba(0,0,0,.3)}.dictionary-item-container{border-bottom:thin solid #ececec}.dictionary-title{padding:5px 10px;cursor:pointer}.dictionary-item-container:last-child{border-bottom-style:none}.dictionary-body{border-top:thin solid #ececec;padding:10px}.dictionary-description{text-align:justify}`],
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
DictionaryListComponent.ctorParameters = () => [
    { type: Store }
];
DictionaryListComponent.propDecorators = {
    metadataIdentifiers: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class DictionaryProgressComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
DictionaryProgressComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'app-dictionary-progress',
                template: `<div class="dictionary-progress">
    <div class="animated-background"></div>
</div>`,
                styles: [`.dictionary-progress div{height:5px;margin-top:3px}@-webkit-keyframes placeHolderShimmer{0%{background-position:-468px 0}100%{background-position:468px 0}}@keyframes placeHolderShimmer{0%{background-position:-468px 0}100%{background-position:468px 0}}.animated-background{-webkit-animation-duration:1.5s;animation-duration:1.5s;-webkit-animation-fill-mode:forwards;animation-fill-mode:forwards;-webkit-animation-iteration-count:infinite;animation-iteration-count:infinite;-webkit-animation-name:placeHolderShimmer;animation-name:placeHolderShimmer;-webkit-animation-timing-function:linear;animation-timing-function:linear;background:#f6f7f8;background:linear-gradient(to right,#eee 8%,#ddd 18%,#eee 33%);background-size:800px 104px;height:100%;position:relative}`]
            },] },
];
/** @nocollapse */
DictionaryProgressComponent.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class DictionaryEffects {
    /**
     * @param {?} actions$
     * @param {?} store
     * @param {?} httpClient
     * @param {?} datePipe
     */
    constructor(actions$, store, httpClient, datePipe) {
        this.actions$ = actions$;
        this.store = store;
        this.httpClient = httpClient;
        this.datePipe = datePipe;
        this.initializeDictionary$ = this.actions$.pipe(ofType(DictionaryActionTypes.InitializeDictionaryMetadata), mergeMap((action) => this.store
            .select(getDictionaryList(action.dictionaryMetadataIdentifiers))
            .pipe(map$1((dictionaryList) => filter(action.dictionaryMetadataIdentifiers, metadataId => !find(dictionaryList, ['id', metadataId]))))), tap(identifiers => {
            /**
                   * Add incoming items to the dictionary list
                   */
            this.store.dispatch(new AddDictionaryMetadataListAction(map(identifiers, id => {
                return {
                    id,
                    name: '',
                    description: '',
                    progress: {
                        loading: true,
                        loadingSucceeded: false,
                        loadingFailed: false
                    }
                };
            })));
            /**
                   * Identify corresponding dictionary items
                   */
            from(identifiers)
                .pipe(mergeMap(identifier => this.httpClient.get(`identifiableObjects/${identifier}.json`, true)))
                .subscribe((metadata) => {
                this.store.dispatch(new UpdateDictionaryMetadataAction(metadata.id, {
                    name: metadata.name,
                    progress: {
                        loading: true,
                        loadingSucceeded: true,
                        loadingFailed: false
                    }
                }));
                if (metadata.href && metadata.href.indexOf('indicator') !== -1) {
                    const /** @type {?} */ indicatorUrl = 'indicators/' +
                        metadata.id +
                        '.json?fields=:all,displayName,id,name,numeratorDescription,' +
                        'denominatorDescription,denominator,numerator,annualized,decimals,indicatorType[name],user[name],' +
                        'attributeValues[value,attribute[name]],indicatorGroups[name,indicators~size],legendSet[name,symbolizer,' +
                        'legends~size],dataSets[name]';
                    this.getIndicatorInfo(indicatorUrl, metadata.id);
                }
                else if (metadata.href &&
                    metadata.href.indexOf('dataElement') !== -1) {
                    const /** @type {?} */ dataElementUrl = 'dataElements/' +
                        metadata.id +
                        '.json?fields=:all,id,name,aggregationType,displayName,' +
                        'categoryCombo[id,name,categories[id,name,categoryOptions[id,name]]],dataSets[:all,!compulsoryDataElementOperands]';
                    this.getDataElementInfo(dataElementUrl, metadata.id);
                }
                else if (metadata.href && metadata.href.indexOf('dataSet') !== -1) {
                    const /** @type {?} */ dataSetUrl = 'dataSets/' +
                        metadata.id +
                        '.json?fields=:all,user[:all],id,name,periodType,shortName,' +
                        'categoryCombo[id,name,categories[id,name,categoryOptions[id,name]]]';
                    this.getDataSetInfo(dataSetUrl, metadata.id);
                }
            });
        }));
    }
    /**
     * @param {?} dataSetUrl
     * @param {?} dataSetId
     * @return {?}
     */
    getDataSetInfo(dataSetUrl, dataSetId) {
        this.httpClient.get(`${dataSetUrl}`, true).subscribe((dataSet) => {
            let /** @type {?} */ dataSetDescription = '<p>' +
                dataSet.name +
                ' of the <strong>' +
                dataSet.formType +
                '</strong> Form created ' +
                'at <strong>' +
                this.datePipe.transform(dataSet.created) +
                ' by ' +
                dataSet.user.name +
                '</strong>';
            if (dataSet.categoryCombo && dataSet.categoryCombo.name !== 'default') {
                dataSetDescription +=
                    '<span> With <strong>' +
                        dataSet.categoryCombo.name +
                        '</strong> Dimension which is divided' +
                        ' into ';
                dataSet.categoryCombo.categories.forEach((category, categoryIndex) => {
                    if (categoryIndex !== 0 &&
                        categoryIndex !== dataSet.categoryCombo.categories.length - 1) {
                        dataSetDescription += ', ';
                    }
                    if (categoryIndex === dataSet.categoryCombo.categories.length - 1 &&
                        dataSet.categoryCombo.categories.length > 1) {
                        dataSetDescription += ' and ';
                    }
                    dataSetDescription += '<strong>';
                    category.categoryOptions.forEach((categoryOption, categoryOptionIndex) => {
                        if (categoryOptionIndex !== 0 &&
                            categoryOptionIndex !== category.categoryOptions.length - 1) {
                            dataSetDescription += ', ';
                        }
                        if (categoryOptionIndex === category.categoryOptions.length - 1 &&
                            category.categoryOptions.length > 1) {
                            dataSetDescription += ' and ';
                        }
                        dataSetDescription += '<span>' + categoryOption.name + '</span>';
                    });
                    dataSetDescription += '</strong>';
                });
                dataSetDescription += '</span>';
            }
            dataSetDescription += '</p>';
            this.store.dispatch(new UpdateDictionaryMetadataAction(dataSetId, {
                description: dataSetDescription,
                progress: {
                    loading: false,
                    loadingSucceeded: true,
                    loadingFailed: false
                }
            }));
        });
    }
    /**
     * @param {?} dataElementUrl
     * @param {?} dataElementId
     * @return {?}
     */
    getDataElementInfo(dataElementUrl, dataElementId) {
        this.httpClient
            .get(`${dataElementUrl}`, true)
            .subscribe((dataElement) => {
            let /** @type {?} */ dataElementDescription = '<p>This ' +
                dataElement.name +
                ' of this method of data aggregation <strong>' +
                dataElement.aggregationType +
                '</strong> created at <strong>' +
                this.datePipe.transform(dataElement.created) +
                '</strong> is only taking <strong>' +
                dataElement.domainType +
                '</strong> data. As the culture of helping user ' +
                'not entering unrecognized data, therefore its only taking <strong>' +
                dataElement.valueType +
                '</strong> values ' +
                'from the user input</p>';
            if (dataElement.categoryCombo.name !== 'default') {
                dataElementDescription +=
                    '<p><strong>' +
                        dataElement.name +
                        '</strong> consists of <strong>' +
                        dataElement.categoryCombo.name +
                        '</strong> category combinations of ';
                dataElement.categoryCombo.categories.forEach((category, index) => {
                    if (index !== 0 &&
                        index !== dataElement.categoryCombo.categories.length - 1) {
                        dataElementDescription += ', ';
                    }
                    if (index === dataElement.categoryCombo.categories.length - 1 &&
                        dataElement.categoryCombo.categories.length > 1) {
                        dataElementDescription += ' and ';
                    }
                    dataElementDescription += '<strong>(';
                    category.categoryOptions.forEach((categoryOption, categoryOptionIndex) => {
                        if (categoryOptionIndex !== 0 &&
                            categoryOptionIndex !== category.categoryOptions.length - 1) {
                            dataElementDescription += ', ';
                        }
                        if (categoryOptionIndex === category.categoryOptions.length - 1 &&
                            category.categoryOptions.length > 1) {
                            dataElementDescription += ' and ';
                        }
                        dataElementDescription +=
                            '<span>' + categoryOption.name + '</span>';
                    });
                    dataElementDescription +=
                        ')</strong> of the <strong>' +
                            category.name +
                            '</strong> category';
                });
                dataElementDescription += '</strong></p>';
                // TODO deal with different version of dhis
                if (dataElement.dataSets && dataElement.dataSets.length > 0) {
                    dataElementDescription +=
                        '<h5>' + dataElement.name + ' Sources</h5>';
                    dataElementDescription +=
                        '<p>More than <strong>' +
                            dataElement.dataSets.length +
                            '</strong> dataset ie ';
                    dataElement.dataSets.forEach((dataSet, dataSetIndex) => {
                        if (dataSetIndex !== 0 &&
                            dataSetIndex !== dataElement.dataSets.length - 1) {
                            dataElementDescription += ', ';
                        }
                        if (dataSetIndex === dataElement.dataSets.length - 1 &&
                            dataElement.dataSets.length > 1) {
                            dataElementDescription += ' and ';
                        }
                        dataElementDescription +=
                            '<strong>' + dataSet.name + '</strong>';
                    });
                    dataElementDescription +=
                        ' use this ' + dataElement.name + ' data element';
                    if (dataElement.dataElementGroups &&
                        dataElement.dataElementGroups.length > 0) {
                        dataElementDescription += ' and it belongs to ';
                        dataElement.dataElementGroups.forEach((dataElementGroup, dataElementGroupIndex) => {
                            if (dataElementGroupIndex !== 0 &&
                                dataElementGroupIndex !==
                                    dataElement.dataElementGroups.length - 1) {
                                dataElementDescription += ', ';
                            }
                            if (dataElementGroupIndex ===
                                dataElement.dataElementGroups.length - 1 &&
                                dataElement.dataElementGroups.length > 1) {
                                dataElementDescription += ' and ';
                            }
                            dataElementDescription +=
                                '<strong>' + dataElementGroup.name + ' Group</strong>';
                        });
                    }
                    dataElementDescription += '</p>';
                }
                this.store.dispatch(new UpdateDictionaryMetadataAction(dataElementId, {
                    description: dataElementDescription,
                    progress: {
                        loading: false,
                        loadingSucceeded: true,
                        loadingFailed: false
                    }
                }));
            }
        });
    }
    /**
     * @param {?} indicatorUrl
     * @param {?} indicatorId
     * @return {?}
     */
    getIndicatorInfo(indicatorUrl, indicatorId) {
        this.httpClient.get(`${indicatorUrl}`, true).subscribe((indicator) => {
            let /** @type {?} */ indicatorDescription = '<p><strong>' +
                indicator.name +
                '</strong> is a <strong>' +
                indicator.indicatorType.name +
                ' </strong> indicator';
            if (indicator.numeratorDescription) {
                indicatorDescription +=
                    '<span> with the numerator described as <strong>' +
                        indicator.numeratorDescription +
                        '</strong></span>';
            }
            if (indicator.denominatorDescription) {
                indicatorDescription +=
                    '<span> and denominator described as <strong>' +
                        indicator.denominatorDescription +
                        '</strong></span>';
            }
            indicatorDescription += '</p>';
            if (indicator.annualized) {
                indicatorDescription +=
                    '<p><span>It’s figure is annualized to support analysis in less than year period ' +
                        '(monthly,quarterly,semi-annually)</span></p>';
            }
            this.store.dispatch(new UpdateDictionaryMetadataAction(indicatorId, {
                description: indicatorDescription,
                progress: {
                    loading: true,
                    loadingSucceeded: true,
                    loadingFailed: false
                }
            }));
            /**
                   * Get numerator expression
                   */
            forkJoin(this.httpClient.get('expressions/description?expression=' +
                encodeURIComponent(indicator.numerator), true), this.httpClient.get('dataSets.json?fields=periodType,id,name,timelyDays,formType,created,expiryDays&' +
                'filter=dataSetElements.dataElement.id:in:[' +
                this.getAvailableDataElements(indicator.numerator) +
                ']&paging=false', true)).subscribe((numeratorResults) => {
                if (numeratorResults[0]) {
                    indicatorDescription +=
                        '<p>Numerator is calculated from <strong>' +
                            numeratorResults[0].description +
                            '</strong>';
                }
                if (numeratorResults[1] && numeratorResults[1].dataSets) {
                    const /** @type {?} */ dataSets = numeratorResults[1].dataSets;
                    if (dataSets.length > 0) {
                        indicatorDescription += ' originating from ';
                    }
                    dataSets.forEach((dataset, index) => {
                        if (index !== 0 && index !== dataSets.length - 1) {
                            indicatorDescription += ', ';
                        }
                        if (index === dataSets.length - 1 && dataSets.length > 1) {
                            indicatorDescription += ' and ';
                        }
                        indicatorDescription +=
                            '<span><strong>' +
                                dataset.name +
                                ',</strong> that is collected <strong>' +
                                dataset.periodType +
                                '</strong> with deadline for submission after <strong>' +
                                dataset.timelyDays +
                                ' days </strong></span>';
                    });
                }
                indicatorDescription += `</p>`;
                this.store.dispatch(new UpdateDictionaryMetadataAction(indicatorId, {
                    description: indicatorDescription,
                    progress: {
                        loading: true,
                        loadingSucceeded: true,
                        loadingFailed: false
                    }
                }));
                /**
                         * Get denominator expression
                         */
                forkJoin(this.httpClient.get('expressions/description?expression=' +
                    encodeURIComponent(indicator.denominator), true), this.httpClient.get('dataSets.json?fields=periodType,id,name,timelyDays,formType,created,expiryDays&' +
                    'filter=dataSetElements.dataElement.id:in:[' +
                    this.getAvailableDataElements(indicator.denominator) +
                    ']&paging=false', true)).subscribe((denominatorResults) => {
                    if (denominatorResults[0]) {
                        indicatorDescription +=
                            '<p>Denominator is calculated from <strong>' +
                                denominatorResults[0].description +
                                '</strong>';
                    }
                    if (denominatorResults[1] && denominatorResults[1].dataSets) {
                        const /** @type {?} */ dataSets = denominatorResults[1].dataSets;
                        if (dataSets.length > 0) {
                            indicatorDescription += ' originating from ';
                        }
                        dataSets.forEach((dataset, index) => {
                            if (index !== 0 && index !== dataSets.length - 1) {
                                indicatorDescription += ', ';
                            }
                            if (index === dataSets.length - 1 && dataSets.length > 1) {
                                indicatorDescription += ' and ';
                            }
                            indicatorDescription +=
                                '<span><strong>' +
                                    dataset.name +
                                    ',</strong> that is collected <strong>' +
                                    dataset.periodType +
                                    '</strong> with deadline for submission after <strong>' +
                                    dataset.timelyDays +
                                    ' days </strong></span>';
                        });
                    }
                    indicatorDescription += `</p>`;
                    /**
                               * Indicator group
                               */
                    if (indicator.indicatorGroups &&
                        indicator.indicatorGroups.length > 0) {
                        indicatorDescription += '<div><p>It belongs to ';
                        indicator.indicatorGroups.forEach((indicatorGroup, index) => {
                            if (index !== 0 &&
                                index !== indicator.indicatorGroups.length - 1) {
                                indicatorDescription += ', ';
                            }
                            if (index === indicator.indicatorGroups.length - 1 &&
                                indicator.indicatorGroups.length > 1) {
                                indicatorDescription += ' and ';
                            }
                            indicatorDescription +=
                                '<span><strong>' +
                                    indicatorGroup.name +
                                    '</strong> with <strong>' +
                                    indicatorGroup.indicators +
                                    '</strong> other related indicators</span>';
                        });
                        indicatorDescription += '</p></div>';
                    }
                    /**
                               * Attribute values
                               */
                    if (indicator.attributeValues &&
                        indicator.attributeValues.length > 0) {
                        indicatorDescription +=
                            '<div><p>Other related details associated with this indicators includes: ';
                        indicator.attributeValues.forEach(attr => {
                            indicatorDescription +=
                                '<span><strong>' +
                                    attr.attribute.name +
                                    ': ' +
                                    attr.value +
                                    '</strong></span>';
                        });
                        indicatorDescription += '</p></div>';
                    }
                    /**
                               * Legend set
                               */
                    if (indicator.legendSet) {
                        indicatorDescription +=
                            '<div><p> It makes use of: <strong>' +
                                indicator.legendSet.name +
                                '</strong> legend' +
                                ' set for analysis with <strong>' +
                                indicator.legendSet.legends +
                                ' Classes </strong>using <strong>' +
                                indicator.legendSet.symbolizer +
                                ' for analysis</strong></p></div>';
                    }
                    /**
                               * User info
                               */
                    if (indicator.user) {
                        indicatorDescription +=
                            '<div><p>This indicator was <strong> first created </strong> in the system on <strong>' +
                                this.datePipe.transform(indicator.created) +
                                '</strong> by <strong>' +
                                indicator.user.name +
                                '</strong></p></div>';
                    }
                    this.store.dispatch(new UpdateDictionaryMetadataAction(indicatorId, {
                        description: indicatorDescription,
                        progress: {
                            loading: false,
                            loadingSucceeded: true,
                            loadingFailed: false
                        }
                    }));
                });
            });
        });
    }
    /**
     * @param {?} data
     * @return {?}
     */
    getAvailableDataElements(data) {
        let /** @type {?} */ dataElementUids = [];
        const /** @type {?} */ separators = [' ', '\\+', '-', '\\(', '\\)', '\\*', '/', ':', '\\?'];
        const /** @type {?} */ numeratorDataElements = data.split(new RegExp(separators.join('|'), 'g'));
        numeratorDataElements.forEach(dataElement => {
            dataElementUids = this.dataElementWithCategoryOptionCheck(dataElement);
        });
        return dataElementUids.join();
    }
    /**
     * @param {?} dataElement
     * @return {?}
     */
    dataElementWithCategoryOptionCheck(dataElement) {
        const /** @type {?} */ uid = [];
        if (dataElement.indexOf('.') >= 1) {
            uid.push(dataElement
                .replace(/#/g, '')
                .replace(/{/g, '')
                .replace(/}/g, '')
                .split('.')[0]);
        }
        else {
            uid.push(dataElement
                .replace(/#/g, '')
                .replace(/{/g, '')
                .replace(/}/g, ''));
        }
        return uid;
    }
}
DictionaryEffects.decorators = [
    { type: Injectable },
];
/** @nocollapse */
DictionaryEffects.ctorParameters = () => [
    { type: Actions },
    { type: Store },
    { type: NgxDhis2HttpClientService },
    { type: DatePipe }
];
__decorate([
    Effect({ dispatch: false }),
    __metadata("design:type", Observable)
], DictionaryEffects.prototype, "initializeDictionary$", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxDhis2DictionaryModule {
}
NgxDhis2DictionaryModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    HttpClientModule,
                    StoreModule.forFeature('dictionary', dictionaryReducer),
                    EffectsModule.forFeature([DictionaryEffects])
                ],
                declarations: [DictionaryListComponent, DictionaryProgressComponent],
                exports: [DictionaryListComponent],
                providers: [DatePipe]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { NgxDhis2DictionaryModule, DictionaryListComponent as ɵd, DictionaryProgressComponent as ɵe, DictionaryEffects as ɵc, dictionaryReducer as ɵb };
