// @hisptz/ngx-dhis2-table/esm2015/lib/components/table-list/table-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { getTableConfiguration } from '../../helpers/index';
export class TableListComponent {
    constructor() {
        this.tableLayers = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.visualizationLayers && this.visualizationLayers.length > 0) {
            this.tableLayers = this.visualizationLayers.map((layer) => {
                return {
                    tableConfiguration: getTableConfiguration(layer.config || {}, layer.layout, this.visualizationType),
                    analyticsObject: layer.analytics
                };
            });
        }
    }
}
TableListComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'ngx-dhis2-table-list',
                template: `<div class="row">
    <div class="col-12" *ngFor="let tableLayer of tableLayers">
        <ngx-dhis2-table-item [analyticsObject]="tableLayer.analyticsObject" [tableConfiguration]="tableLayer.tableConfiguration">
        </ngx-dhis2-table-item>
    </div>
</div>`,
                styles: [``]
            },] },
];
/** @nocollapse */
TableListComponent.ctorParameters = () => [];
TableListComponent.propDecorators = {
    visualizationLayers: [{ type: Input }],
    visualizationType: [{ type: Input }]
};
function TableListComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    TableListComponent.prototype.visualizationLayers;
    /** @type {?} */
    TableListComponent.prototype.visualizationType;
    /** @type {?} */
    TableListComponent.prototype.tableLayers;
}
