// @hisptz/ngx-dhis2-dictionary/esm5/lib/components/dictionary-list/dictionary-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { Store } from '@ngrx/store';
import { getDictionaryList } from '../../store/selectors/dictionary.selectors';
import { InitializeDictionaryMetadataAction } from '../../store/actions/dictionary.actions';
var DictionaryListComponent = /** @class */ (function () {
    function DictionaryListComponent(store) {
        this.store = store;
        this.activeItem = 0;
    }
    /**
     * @return {?}
     */
    DictionaryListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.metadataIdentifiers.length > 0) {
            this.store.dispatch(new InitializeDictionaryMetadataAction(this.metadataIdentifiers));
            this.dictionaryList$ = this.store.select(getDictionaryList(this.metadataIdentifiers));
        }
    };
    /**
     * @param {?} index
     * @param {?} e
     * @return {?}
     */
    DictionaryListComponent.prototype.setActiveItem = /**
     * @param {?} index
     * @param {?} e
     * @return {?}
     */
    function (index, e) {
        e.stopPropagation();
        if (this.activeItem === index) {
            this.activeItem = -1;
        }
        else {
            this.activeItem = index;
        }
    };
    DictionaryListComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'ngx-dhis2-dictionary-list',
                    template: "<div class=\"dictionary-list-container\">\n  <div\n    class=\"dictionary-item-container\"\n    *ngFor=\"let dictionaryItem of dictionaryList$ | async; let i = index\"\n  >\n    <div\n      class=\"dictionary-title\"\n      (click)=\"setActiveItem(i, $event)\"\n    >\n      <span>{{ dictionaryItem.name }}</span>\n      <div *ngIf=\"!dictionaryItem.name\">\n        <app-dictionary-progress></app-dictionary-progress>\n      </div>\n    </div>\n    <div\n      class=\"dictionary-body\"\n      *ngIf=\"activeItem === i\"\n    >\n      <div class=\"dictionary-description\" [innerHtml]=\"dictionaryItem.description\"></div>\n      <div *ngIf=\"dictionaryItem.progress.loading\">\n        <app-dictionary-progress></app-dictionary-progress>\n      </div>\n    </div>\n  </div>\n</div>\n",
                    styles: [".dictionary-list-container{box-shadow:0 0 3px rgba(0,0,0,.3)}.dictionary-item-container{border-bottom:thin solid #ececec}.dictionary-title{padding:5px 10px;cursor:pointer}.dictionary-item-container:last-child{border-bottom-style:none}.dictionary-body{border-top:thin solid #ececec;padding:10px}.dictionary-description{text-align:justify}"],
                    changeDetection: ChangeDetectionStrategy.OnPush
                },] },
    ];
    /** @nocollapse */
    DictionaryListComponent.ctorParameters = function () { return [
        { type: Store }
    ]; };
    DictionaryListComponent.propDecorators = {
        metadataIdentifiers: [{ type: Input }]
    };
    return DictionaryListComponent;
}());
export { DictionaryListComponent };
function DictionaryListComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DictionaryListComponent.prototype.metadataIdentifiers;
    /** @type {?} */
    DictionaryListComponent.prototype.dictionaryList$;
    /** @type {?} */
    DictionaryListComponent.prototype.activeItem;
    /** @type {?} */
    DictionaryListComponent.prototype.store;
}
