// @hisptz/ngx-dhis2-dictionary/esm2015/lib/store/actions/dictionary.actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {string} */
const DictionaryActionTypes = {
    InitializeDictionaryMetadata: '[Dictionary] initialize incoming metadata',
    AddDictionaryMetadataList: '[Dictionary] add dictionary metadata list',
    UpdateDictionaryMetadata: '[Dictionary] update dictionary metadata list',
};
export { DictionaryActionTypes };
export class InitializeDictionaryMetadataAction {
    /**
     * @param {?} dictionaryMetadataIdentifiers
     */
    constructor(dictionaryMetadataIdentifiers) {
        this.dictionaryMetadataIdentifiers = dictionaryMetadataIdentifiers;
        this.type = DictionaryActionTypes.InitializeDictionaryMetadata;
    }
}
function InitializeDictionaryMetadataAction_tsickle_Closure_declarations() {
    /** @type {?} */
    InitializeDictionaryMetadataAction.prototype.type;
    /** @type {?} */
    InitializeDictionaryMetadataAction.prototype.dictionaryMetadataIdentifiers;
}
export class AddDictionaryMetadataListAction {
    /**
     * @param {?} dictionaryMetadataList
     */
    constructor(dictionaryMetadataList) {
        this.dictionaryMetadataList = dictionaryMetadataList;
        this.type = DictionaryActionTypes.AddDictionaryMetadataList;
    }
}
function AddDictionaryMetadataListAction_tsickle_Closure_declarations() {
    /** @type {?} */
    AddDictionaryMetadataListAction.prototype.type;
    /** @type {?} */
    AddDictionaryMetadataListAction.prototype.dictionaryMetadataList;
}
export class UpdateDictionaryMetadataAction {
    /**
     * @param {?} dictionaryMetadataId
     * @param {?} changes
     */
    constructor(dictionaryMetadataId, changes) {
        this.dictionaryMetadataId = dictionaryMetadataId;
        this.changes = changes;
        this.type = DictionaryActionTypes.UpdateDictionaryMetadata;
    }
}
function UpdateDictionaryMetadataAction_tsickle_Closure_declarations() {
    /** @type {?} */
    UpdateDictionaryMetadataAction.prototype.type;
    /** @type {?} */
    UpdateDictionaryMetadataAction.prototype.dictionaryMetadataId;
    /** @type {?} */
    UpdateDictionaryMetadataAction.prototype.changes;
}
